src/gfn/env.py

```diff
@@ -56,14 +73,89 @@ def __init__(
         self.preprocessor = preprocessor
         self.is_discrete = False
 
+    def states_from_tensor(self, tensor: Tensor):
+        """Wraps the supplied Tensor in a States instance."""
+        return self.States(tensor)
+
+    def states_from_batch_shape(self, batch_shape: Tuple):
+        """Returns a batch of s0 states with a given batch_shape."""
+        return self.States.from_batch_shape(batch_shape)
+
+    def actions_from_tensor(self, tensor: Tensor):
+        """Wraps the supplied Tensor an an Actions instance."""
+        return self.Actions(tensor)
+
+    def actions_from_batch_shape(self, batch_shape: Tuple):
+        """Returns a batch of dummy actions with the supplied batch_shape."""
+        return self.Actions.make_dummy_actions(batch_shape)
+
+    # To be implemented by the User.
     @abstractmethod
-    def make_States_class(self) -> type[States]:
-        """Returns a class that inherits from States and implements the environment-specific methods."""
+    def step(
+        self, states: States, actions: Actions
+    ) -> TT["batch_shape", "state_shape", torch.float]:
+        """Function that takes a batch of states and actions and returns a batch of next
+        states. Does not need to check whether the actions are valid or the states are sink states.
+        """
+
+    @abstractmethod
+    def backward_step(  # TODO: rename to backward_step, other method becomes _backward_step.
+        self, states: States, actions: Actions
+    ) -> TT["batch_shape", "state_shape", torch.float]:
+        """Function that takes a batch of states and actions and returns a batch of previous
+        states. Does not need to check whether the actions are valid or the states are sink states.
+        """
 
     @abstractmethod
+    def is_action_valid(
+        self,
+        states: States,
+        actions: Actions,
+        backward: bool = False,
+    ) -> bool:
+        """Returns True if the actions are valid in the given states."""
+
+    def make_random_states_tensor(self, batch_shape: Tuple) -> Tensor:
+        """Optional method inherited by all States instances to emit a random tensor."""
+        raise NotImplementedError
+
+    # Optionally implemented by the user when advanced functionality is required.
+    def make_States_class(self) -> type[States]:
+        """The default States class factory for all Environments.
+
+        Returns a class that inherits from States and implements assumed methods.
+        The make_States_class method should be overwritten to achieve more
+        environment-specific States functionality.
+        """
+        env = self
+
+        class DefaultEnvState(States):
+            """Defines a States class for this environment."""
+
+            state_shape = env.state_shape
+            s0 = env.s0
+            sf = env.sf
+            make_random_states_tensor = env.make_random_states_tensor
+
+        return DefaultEnvState
+
     def make_Actions_class(self) -> type[Actions]:
```

Comment:
Same comment here - I know that the class is `Actions` but I still would advocate that this method should be `make_actions_class` to be more inline with PEP 8.